model/connect/Database.php

```diff
@@ -489,6 +489,28 @@ public function affectedRows() {
 	 */
 	abstract public function supportsTransactions();
 
+	/**
+	 * Invoke $callback within a transaction
+	 *
+	 * @param callable $callback Callback to run
+	 * @param callable $errorCallback Optional callback to run after rolling back transaction.
+	 * @param bool|string $transactionMode Optional transaction mode to use
+	 * @throws Exception
+	 */
+	public function withTransaction($callback, $errorCallback = null, $transactionMode = false) {
```

Comment:
As is, it's quite awkward for the caller to wrap a callback in a withTransaction block, but if the DB doesn't support transactions just use the callback directly.

We probably want an $errorIfTransactionsUnsupported parameter defaulting to false, and to check supportsTransactions on each step.